coreneuron/sim/fadvance_core.cpp

```diff
@@ -293,28 +293,29 @@ void nrncore2nrn_send_values(NrnThread* nth) {
 
     TrajectoryRequests* tr = nth->trajec_requests;
     if (tr) {
-        // Note that this is rather inefficient: we generate one `acc update
-        // self` call for each `double` value (voltage, membrane current,
-        // mechanism property, ...) that is being recorded, even though in most
-        // cases these values will actually fall in a small number of contiguous
-        // ranges in memory. A much better solution, which should be implemented
-        // soon, would be to gather and buffer these values timestep-by-timestep
-        // in GPU memory, and only upload the results to the host at the end of
-        // the simulation. See also: https://github.com/BlueBrain/CoreNeuron/issues/611
         if (tr->varrays) {  // full trajectories into Vector data
-            double** va = tr->varrays;
             int vs = tr->vsize++;
+            // make sure we do not overflow the `varrays` buffers
             assert(vs < tr->bsize);
-            for (int i = 0; i < tr->n_trajec; ++i) {
-                double* gather_i = tr->gather[i];
-                // clang-format off
 
-                #pragma acc update self(gather_i[0:1]) if(nth->compute_gpu)
-                // clang-format on
-                va[i][vs] = *gather_i;
+            // clang-format off
+
+            #pragma acc parallel loop present(tr[0:1]) if(nth->compute_gpu) async(nth->stream_id)
+            // clang-format on
+            for (int i = 0; i < tr->n_trajec; ++i) {
+                tr->varrays[i][vs] = *tr->gather[i];
             }
         } else if (tr->scatter) {  // scatter to NEURON and notify each step.
             nrn_assert(nrn2core_trajectory_values_);
+            // Note that this is rather inefficient: we generate one `acc update
+            // self` call for each `double` value (voltage, membrane current,
+            // mechanism property, ...) that is being recorded, even though in most
+            // cases these values will actually fall in a small number of contiguous
+            // ranges in memory. A better solution, if the performance of this
+            // branch becomes limiting, might be to offload this loop to the
+            // device and populate some `scatter_values` array there and copy it
+            // back with a single transfer. See also:
+            // https://github.com/BlueBrain/CoreNeuron/issues/611
             for (int i = 0; i < tr->n_trajec; ++i) {
```

Comment:
So - this part is still inefficient because we don't know "ranges" to update in one go and these needs to be sent back in every timestep, right?

Reply:
Yes.

Not knowing which values are contiguous means we have `n_trajec` updates per time step instead of something like `n_properties_being_recorded` per time step (something like `n_trajec / n_sections` if variables are recorded on all sections/cells). We need to have **some** updates every timestep because apparently we have to call the callback every timestep.

Looking at this again this morning I think I need to re-add the `acc wait` I moved out of `nrn_fixed_step_lastpart` to be just before this loop, otherwise it might not pick up the correct values of mechanism data that is modified in `nrn_state` methods (inside `nonvint`).

I should also add that this branch is being modified "blind"...I am not aware of any test coverage of the `else if (tr->scatter) { ... }` part.

Reply:
> Looking at this again this morning I think I need to re-add the acc wait I moved out of nrn_fixed_step_lastpart to be just before this loop, otherwise it might not pick up the correct values of mechanism data that is modified in nrn_state methods (inside nonvint).

For the first loop / if block :
>  if (tr->varrays) {  // full trajectories into Vector data
> ...
>         tr->varrays[i][vs] = *tr->gather[i];

Yesterday it came to my mind I thought: `nrn_state` are updating data on GPU and the loop is copying that data into `tr->varrays[i][vs]` which is also on GPU. As the kernels are launched in same stream `async(nth->stream_id)`, they will be executed in order and hence will pick-up correct values. So there is no need for explicit wait ?

Reply:
Indeed, I meant I needed to re-add the `acc wait` for the `else if(tr->scatter)` block: https://github.com/BlueBrain/CoreNeuron/blob/14b083632a92654954014efd7f28c4eb3d0294fc/coreneuron/sim/fadvance_core.cpp#L324-L333